File: app/src/main/java/at/bitfire/icsdroid/ui/TitleColorFragment.java
/*
 * Copyright (c) Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.icsdroid.ui;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;

import at.bitfire.icsdroid.R;
import yuku.ambilwarna.AmbilWarnaDialog;

public class TitleColorFragment extends Fragment implements TextWatcher {
    public static final String
            ARG_URL = "url",
            ARG_TITLE = "title",
            ARG_COLOR = "color";

    private String url;

    private EditText editTitle;
    String title;

    private ColorButton colorButton;
    int color = 0xff2F80C7;

    private OnChangeListener listener;


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle inState) {
        View v = inflater.inflate(R.layout.calendar_title_color, container, false);

        if (inState == null) {
            Bundle args = getArguments();
            url = args.getString(ARG_URL);
            title = args.getString(ARG_TITLE);
            color = args.getInt(ARG_COLOR);
        } else {
            url = inState.getString(ARG_URL);
            title = inState.getString(ARG_TITLE);
            color = inState.getInt(ARG_COLOR);
        }

        TextView textURL = (TextView)v.findViewById(R.id.url);
        textURL.setText(url);

        editTitle = (EditText) v.findViewById(R.id.title);
        setTitle(title);
        editTitle.addTextChangedListener(this);

        colorButton = (ColorButton) v.findViewById(R.id.color);
        colorButton.setColor(color);
        colorButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new AmbilWarnaDialog(getActivity(), color, new AmbilWarnaDialog.OnAmbilWarnaListener() {
                    @Override
                    public void onCancel(AmbilWarnaDialog ambilWarnaDialog) {
                    }

                    @Override
                    public void onOk(AmbilWarnaDialog ambilWarnaDialog, int newColor) {
                        colorButton.setColor(color = 0xff000000 | newColor);
                        notifyListener();
                    }
                }).show();
            }
        });

        return v;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(ARG_URL, url);
        outState.putString(ARG_TITLE, title);
        outState.putInt(ARG_COLOR, color);
    }


    public void setTitle(String title) {
        editTitle.setText(this.title = title);
    }


    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
    }

    @Override
    public void afterTextChanged(Editable s) {
        title = editTitle.getText().toString();
        notifyListener();
    }


    public interface OnChangeListener {
        void onChangeTitleColor(String title, int color);
    }

    public void setOnChangeListener(OnChangeListener onChangeListener) {
        listener = onChangeListener;
    }

    void notifyListener() {
        if (listener != null)
            listener.onChangeTitleColor(title, color);
    }
}

File: app/src/main/java/at/bitfire/icsdroid/ui/TitleColorFragment.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/

package at.bitfire.icsdroid.ui

import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.CompoundButton.OnCheckedChangeListener
import android.widget.EditText
import android.widget.TextView
import androidx.core.widget.addTextChangedListener
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import at.bitfire.icsdroid.R
import at.bitfire.icsdroid.databinding.TitleColorBinding
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.switchmaterial.SwitchMaterial
import org.joda.time.Minutes
import org.joda.time.format.PeriodFormat

class TitleColorFragment : Fragment() {

    private val model by activityViewModels<TitleColorModel>()

    private lateinit var binding: TitleColorBinding

    /**
     * Provides an observer for the default alarm fields.
     * @param switch The switch view that updates the currently stored minutes.
     * @param textView The viewer for the current value of the stored minutes.
     * @param onValueChanged Gets called when the value of the stored minutes must be updated.
     * Usually contains a call to a ViewModel to update the stored state.
     */
    private fun defaultAlarmObserver(
        switch: SwitchMaterial,
        textView: TextView,
        onValueChanged: (Long?) -> Unit,
    ): Observer<Long> {
        val checkboxCheckedChanged = OnCheckedChangeListener { _, checked ->
            if (!checked) {
                onValueChanged(null)
                return@OnCheckedChangeListener
            }

            val editText = EditText(requireContext()).apply {
                setHint(R.string.default_alarm_dialog_hint)
                isSingleLine = true
                maxLines = 1
                imeOptions = EditorInfo.IME_ACTION_DONE
                inputType = InputType.TYPE_CLASS_NUMBER

                addTextChangedListener { txt ->
                    val text = txt?.toString()
                    val num = text?.toLongOrNull()
                    error = if (text == null || text.isBlank() || num == null)
                        getString(R.string.default_alarm_dialog_error)
                    else
                        null
                }
            }
            MaterialAlertDialogBuilder(requireContext())
                .setTitle(R.string.default_alarm_dialog_title)
                .setMessage(R.string.default_alarm_dialog_message)
                .setView(editText)
                .setPositiveButton(R.string.default_alarm_dialog_set) { dialog, _ ->
                    if (editText.error == null) {
                        onValueChanged(editText.text?.toString()?.toLongOrNull())
                        dialog.dismiss()
                    }
                }
                .setOnCancelListener {
                    switch.isChecked = false
                }
                .create()
                .show()
        }

        return Observer<Long> { min: Long? ->
            switch.isChecked = min != null
            // We add the listener once the switch has an initial value
            switch.setOnCheckedChangeListener(checkboxCheckedChanged)

            if (min == null) {
                textView.visibility = View.GONE
            } else {
                val alarmPeriodText = PeriodFormat.wordBased().print(Minutes.minutes(min.toInt()))
                textView.text = getString(R.string.add_calendar_alarms_default_description, alarmPeriodText)
                textView.visibility = View.VISIBLE
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, inState: Bundle?): View {
        binding = TitleColorBinding.inflate(inflater, container, false)
        binding.lifecycleOwner = this
        binding.model = model

        model.defaultAlarmMinutes.observe(
            viewLifecycleOwner,
            defaultAlarmObserver(binding.defaultAlarmSwitch, binding.defaultAlarmText) {
                model.defaultAlarmMinutes.postValue(it)
            }
        )
        model.defaultAllDayAlarmMinutes.observe(
            viewLifecycleOwner,
            defaultAlarmObserver(binding.defaultAlarmAllDaySwitch, binding.defaultAlarmAllDayText) {
                model.defaultAllDayAlarmMinutes.postValue(it)
            }
        )

        val colorPickerContract = registerForActivityResult(ColorPickerActivity.Contract()) { color ->
            model.color.postValue(color)
        }
        binding.color.setOnClickListener {
            colorPickerContract.launch(model.color.value)
        }

        return binding.root
    }

    class TitleColorModel : ViewModel() {
        var url = MutableLiveData<String>()

        var originalTitle: String? = null
        val title = MutableLiveData<String>()

        var originalColor: Int? = null
        val color = MutableLiveData<Int>()

        var originalIgnoreAlerts: Boolean? = null
        val ignoreAlerts = MutableLiveData<Boolean>()

        var originalDefaultAlarmMinutes: Long? = null
        val defaultAlarmMinutes = MutableLiveData<Long>()

        var originalDefaultAllDayAlarmMinutes: Long? = null
        val defaultAllDayAlarmMinutes = MutableLiveData<Long>()

        fun dirty(): Boolean = originalTitle != title.value || originalColor != color.value || originalIgnoreAlerts != ignoreAlerts.value ||
                originalDefaultAlarmMinutes != defaultAlarmMinutes.value || originalDefaultAllDayAlarmMinutes != defaultAllDayAlarmMinutes.value
    }

}
